### PR TITLE
wayland: remove top level decorator and other wlrootisms.

### LINKED ISSUE #2126: wayland: mutter output detect deadlocks.
detecting current monitors on mutter's wayland compositor deadlocks.

### PR DESCRIPTION
**Pull request by james-lawrence**
resolves obvious issues with running on a mutter wayland compositor.

fundamentally this allows druid applications to render on mutter using wayland.

fixes #2126

### CODE REVIEW

**Comment by james-lawrence** [comment]
@Maan2003 this PR hammers out the remaining bugs I had for my own application (layershell reinitialization issues, slowness during application initialization) and gets druid rendering on mutter (gnome's wayland compositor).

I've covered what I think are the remaining issues in #1956 and filed issues. should mostly be minor issues and just unimplemented internal apis. most of the issues remaining I suspect also impact the X11 backend. good news everybody!

**Comment by james-lawrence** [comment]
@derekdreery your initial work was super helpful. thanks, we wouldn't be at this point without it.
@Maan2003 thanks for all your patience and time spent reviewing.

**Comment by HeavyRain266** [comment]
I think you should make this as optional features instead, xdg-decorations protocol is official extension used by other compositors, not just wlroots one. GNOME refuses to support it because they don't provide any server-side decorations (and uses deprecated kde decorations in gtk). If you completely remove this, then it will cause issues on e.g. KDE or any other compositor which uses custom server-side titlebars.

**Comment by i509VCB** [comment]
Per this https://github.com/linebender/druid/blob/master/druid-shell/src/backend/wayland/application.rs#L204-L207

You wrongly assert xdg_decoration is available.

The proper fix here is to gracefully handle said protocol not being available and draw csd when the protocol isn't available.

**Comment by i509VCB** [comment]
Oh and also the layer shell is not available on mutter, a runtime test for layers will be needed.

**Comment by PolyMeilex** [comment]
As previously mentioned, XDG Decoration is an official protocol, it should not be removed. It's lack should just be gracefully handled like any other non-core protocol. (Maybe exept xdg-shell). 

Your layer shell change is correct tho, xdg decor should be handled the same way.

**Comment by derekdreery** [comment]
@PolyMeilex Could you write a bullet point list of the changes you think should be made to this PR?

File: druid-shell/src/backend/wayland/surfaces/toplevel.rs [hunk 1]
```diff
@@ -97,25 +91,16 @@ impl Surface {
...
         let inner = Inner {
             wl_surface,
             xdg_toplevel,
             xdg_surface,
-            zxdg_toplevel_decoration_v1,
         };
 
         inner
-            .zxdg_toplevel_decoration_v1
-            .set_mode(toplevel_decorations::Mode::ServerSide);
```
[the remark this answers is not in the document]
Comment (PolyMeilex) [reply druid-shell/src/backend/wayland/surfaces/toplevel.rs:113]: Also Druid has a property called `show_titlebar`, and if I understand docs correctly:  `Set whether the window should have a titlebar and decorations.` the property should be passed down to this point and used to decide if we want SSD

**Review by PolyMeilex — COMMENTED** [review]
> Could you write a bullet point list of the changes you think should be made to this PR?

Basically those review comments, besides that everything looks good.

XDG decor gives us free decoration whenever user forcefully unfullscreen the window, so removing this small bit of code would be a big regression.

File: druid-shell/src/backend/wayland/surfaces/toplevel.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/surfaces/toplevel.rs:113]: unrelated to this changeset outside of removing the logic for the time being. it wasn't correct to begin with as you've noted.

**Comment by james-lawrence** [comment]
People are welcome to implement CSD/SSD if they so choose. its not the point of this PR. which is about layershell fixes and toplevel initial config for application startup.

but outside of that scope i'm not going to make additional changes. this PR is feature complete from my perspective.

feel free to implement the features you feel are missing.

File: druid-shell/src/backend/wayland/application.rs [hunk 2]
```diff
@@ -201,14 +199,17 @@ impl Application {
         let env = display::new(dispatcher)?;
         display::print(&env.registry);
 
-        let zxdg_decoration_manager_v1 = env
-            .registry
-            .instantiate_exact::<ZxdgDecorationManagerV1>(1)
-            .map_err(|e| Error::global("zxdg_decoration_manager_v1", 1, e))?;
```
[the remark this answers is not in the document]
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/application.rs:207]: delete coded doesn't need a map or else ;)

File: druid-shell/src/backend/wayland/surfaces/layershell.rs [hunk 3]
```diff
@@ -194,7 +194,7 @@ impl Surface {
     ) -> Self {
         let compositor = CompositorHandle::new(c);
         let wl_surface = surface::Surface::new(compositor.clone(), handler, kurbo::Size::ZERO);
-        let ls_surface = compositor.zwlr_layershell_v1().get_layer_surface(
+        let ls_surface = compositor.zwlr_layershell_v1().unwrap().get_layer_surface(
```
[the remark this answers is not in the document]
Comment (james-lawrence) [reply druid-shell/src/backend/wayland/surfaces/layershell.rs:197]: generally I agree but unnecessary atm. if a application requests layershell and its not available its going to error out anyways. not really concerned if its a hard crash or not.

**Comment by james-lawrence** [comment]
@HeavyRain266 

> If you completely remove this, then it will cause issues on e.g. KDE or any other compositor which uses custom server-side titlebars.

yes. it was already causing issues which is why I removed it. it was incorrectly implemented logic. the goal wasn't to fix it for every platform; the goal is to clean it out so someone implement it correctly later without confusion from pre-existing code.

**Comment by PolyMeilex** [comment]
Oh sorry, just by reading the PR and the linked issue I wrongful assumed that this PR is meant to fix something, but the point is just removal, not a fix. That was most likely discussed somewhere else, my bad. Have a nice day.

**Comment by james-lawrence** [comment]
@PolyMeilex understandable; title is confusing given the decorations being removed (since they're a general api as noted). basically the top level decorations predate my work. there where a bunch of partially implemented areas. I've fixed up most of them; but CSD and SSD are of little interest for my work and this allows the application to render its contents on mutter and wlroots. I'm assuming it'd also render on KDE.

i decided it was better to let someone start with a clean slate if they decided that area interested them vs having a broken partial implementation lying around causing issues and confusion.

**Comment by james-lawrence** [comment]
@PolyMeilex it does fix an issue; right now in main there is a 1sish delay during startup and some issues with restoring layershell when a monitor restarts. the CSD/SSD stuff also caused bugs due to not being implemented properly.

@Maan2003 is there any interest in pulling this in if I resolve the conflict?

**Comment by PolyMeilex** [comment]
I don't necessarily agree with the mindset, but I'm fine with the change btw, at least druid will start up without the need to edit the source code.
So I'm in favor of the merge.

**Review by Maan2003 — APPROVED** [review]
LGTM!

Here is my understanding:

We were using xdg_decoration, which is not supported by mutter. The implementation had bugs which causes delay during startup.
So for now, we are just removing the top level decorator.

I will be able to test in 2-3 days.

**Comment by PolyMeilex** [comment]
> LGTM!
> 
> Here is my understanding:
> 
> We were using xdg_decoration, which is not supported by mutter. The implementation had bugs which causes delay during startup.
> So for now, we are just removing the top level decorator.
> 
> I will be able to test in 2-3 days.

Those are two unrelated changes as far as I know.

Binding a protocol or even version that is not advertised by the compositor is treated as error in Druid. Therefore druid does not run at all on reference implementation Weston and Gnome's Mutter. As it expects optional xdg-decoration protocol and nonstandard layer shell. This PR removes xdg decor and makes layer shell optional and binds it only if available.

Second part of the PR fixes toplevel initialization speed, by changing the order a bit. And I think it also fixed a deadlock in WlOutput code for me.

**Comment by derekdreery** [comment]
People on this thread seem to know a lot more about Wayland than I do, and so I won't comment on the details, but it does seem to me that whichever way that discussion goes this PR is an improvement over what is already there, and so we should get it merged ASAP.

**Comment by james-lawrence** [comment]
@Maan2003 all cleaned up and ready to go. thanks as always.

**Pull request merged**